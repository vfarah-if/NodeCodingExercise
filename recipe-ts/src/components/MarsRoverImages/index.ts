
import MarsRoverImages from './MarsRoverImages';
export default MarsRoverImages;    
