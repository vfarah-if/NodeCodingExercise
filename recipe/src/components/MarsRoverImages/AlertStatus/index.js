
import AlertStatus from './AlertStatus';
export default AlertStatus;    
