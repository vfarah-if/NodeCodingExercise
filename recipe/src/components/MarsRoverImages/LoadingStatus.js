import React from 'react';

export const LoadingStatus = ({ isLoading, displayValue }) => {
    return isLoading ? <p>{displayValue || 'Loading ...'}</p> : null;
}

export default LoadingStatus;
