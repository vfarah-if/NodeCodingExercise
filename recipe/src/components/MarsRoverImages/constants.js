export const ROVER_CAMERAS = {
	spirit: [
		{
			abbrev: 'NAVCAM',
			name: 'Navigation Camera',
		},
		{
			abbrev: 'PANCAM',
			name: 'Panoramic Camera',
		},
	],
	curiosity: [
		{
			abbrev: 'FHAZ',
			name: 'Front Hazard Avoidance Camera',
		},
		{
			abbrev: 'RHAZ',
			name: 'Rear Hazard Avoidance Camera',
		},
		{
			abbrev: 'MAST',
			name: 'Mast Camera',
		},
		{
			abbrev: 'CHEMCAM',
			name: 'Chemistry and Camera Complex',
		},
		{
			abbrev: 'NAVCAM',
			name: 'Navigation Camera',
		},
	],
	opportunity: [
		{
			abbrev: 'NAVCAM',
			name: 'Navigation Camera',
		},
		{
			abbrev: 'PANCAM',
			name: 'Panoramic Camera',
		},
	],
};