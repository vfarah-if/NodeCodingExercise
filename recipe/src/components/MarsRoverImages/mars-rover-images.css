.Mars-Rover-Images {
	text-align: center;
}

.Mars-Rover-logo {
	height: 15vmin;
	pointer-events: none;
}

.Mars-Rover-header {
	background-color: #282c34;
	min-height: 100vh;
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	font-size: calc(10px + 2vmin);
	color: white;
}

.Mars-Rover-link {
	color: #61dafb;
}

.thumbnail {
	/* border: 1px solid #ddd; */
	border-radius: 4px;
	padding: 10px;
	width: 250px;
}

.thumbnail:hover {
	box-shadow: 0 0 2px 1px rgba(0, 140, 186, 0.5);
}

.cards {
	display: grid;
	grid-gap: 20px;
	grid-template-columns: repeat(4, 1fr);
}

.card {
	position: relative;
	box-shadow: 0 3px 6px rgba(0, 0, 0, 0.5);
}

.card:hover {
	cursor: pointer;
}

.card img:not(:first-of-type) {
	position: absolute;
	top: 0;
	right: 0;
	bottom: 0;
	left: 0;
	opacity: 0;
}

.card img.is-visible {
	opacity: 1;
}

ul {
	list-style: none;
}