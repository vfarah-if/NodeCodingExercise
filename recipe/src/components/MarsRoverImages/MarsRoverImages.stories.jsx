import React, { Fragment } from 'react';
import MarsRoverImages from './MarsRoverImages';

// eslint-disable-next-line import/no-anonymous-default-export
export default {
	title: 'Component/MarsRoverImages',
	component: MarsRoverImages,
};

const MarsRoverImagesTemplate = (args) => (
	<Fragment>
		<MarsRoverImages {...args} />
	</Fragment>
);

export const MarsRoverDefaultStory = MarsRoverImagesTemplate.bind({});
MarsRoverDefaultStory.args = {};

export const MarsRoverWithPresetsStory = MarsRoverImagesTemplate.bind({});
MarsRoverWithPresetsStory.args = { roverType: 'spirit', cameraType:'NAVCAM' };
