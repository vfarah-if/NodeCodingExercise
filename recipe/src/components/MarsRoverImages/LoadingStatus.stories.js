import React from 'react';
import LoadingStatus from './LoadingStatus';

// eslint-disable-next-line import/no-anonymous-default-export
export default {
	title: 'Example/LoadingStatus',
	component: LoadingStatus,
};

const LoadingStatusTemplate = (args) => <LoadingStatus {...args} />;

export const DefaultLoadingStatusStory = LoadingStatusTemplate.bind({});
DefaultLoadingStatusStory.args = { isLoading: true };

export const IncorporateCustomDisplayNameStory = LoadingStatusTemplate.bind({});
IncorporateCustomDisplayNameStory.args = {
	isLoading: true,
	displayValue: 'Poor mans loader ...',
};
