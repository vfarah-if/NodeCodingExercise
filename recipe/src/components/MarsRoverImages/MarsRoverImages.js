import React, { useState, useEffect } from 'react';

import LoadingStatus from '../LoadingStatus';
import AlertStatus from '../AlertStatus';
import ShinySelect from '../ShinySelect';
import Thumbnail from '../Thumbnail';
import Cards from '../Cards';
import Card from '../Card';
import config from '../../config';
import { ROVER_OPTIONS, ROVER_CAMERA_OPTIONS } from './constants';
import nasaLogo from '../../assets/nasa-logo-300x250.png';
import noImage from '../../assets/no-photo-400X300.png';
import './mars-rover-images.css';
const { NASAPIKey } = config;

export default function MarsRoverImages() {
	const [roverChosen, setRoverChosen] = useState('curiosity');
	const [cameraChosen, setCameraChosen] = useState('');
	const [error, setError] = useState(null);
	const [successMessage, setSuccessMessage] = useState(null);
	const [isLoading, setIsLoading] = useState(true);
	const [photos, setPhotos] = useState([]);
	const fetchMarsRoverPhotos = async () => {
		try {
			setIsLoading(true);
			if (roverChosen && cameraChosen && cameraChosen !== '') {
				const response = await fetch(
					`https://api.nasa.gov/mars-photos/api/v1/rovers/${roverChosen}/photos?sol=100&camera=${cameraChosen}&api_key=${NASAPIKey}`
				);
				const { photos } = await response.json();
				setPhotos(photos);
				setSuccessMessage(
					`Succeeded retrieving '${photos.length}' photos!`
				);
			} else {
				setPhotos([]);
			}
			return photos;
		} catch (error) {
			setError(error);
		} finally {
			setIsLoading(false);
		}
	};

	useEffect(() => {
		fetchMarsRoverPhotos();
		// eslint-disable-next-line
	}, [roverChosen, cameraChosen]);

	const handleRoverSelection = (event) => {
		setRoverChosen(event.target.value);
		setCameraChosen('');
		setSuccessMessage(null);
	};

	const handleCameraSelection = (event) => {
		setCameraChosen(event.target.value);
	};

	const hasNoPhotos = !photos?.length;

	return (
		<div className="Mars-Rover-Images">
			<div className="App-intro">
				<img
					src={nasaLogo}
					className="Mars-Rover-logo"
					alt="logo"
				></img>
			</div>
			<span>Data provided by NASA.</span>
			<br />
			<span>
				Each photo was taken on the rover's 100th sol (day) on Mars.
			</span>
			<br />
			<br />
			<ShinySelect
				displayName="Choose a Mars rover:"
				value={roverChosen}
				optionsList={ROVER_OPTIONS}
				onChange={handleRoverSelection}
			/>
			<br />
			<ShinySelect
				displayName="Choose a camera:"
				value={cameraChosen}
				optionsList={ROVER_CAMERA_OPTIONS}
				onChange={handleCameraSelection}
			/>
			<div>
				<LoadingStatus
					isLoading={isLoading}
					displayValue="Loading images ..."
				></LoadingStatus>
				<br />
				<AlertStatus alertType="success" message={successMessage} />
			</div>
			{hasNoPhotos ? (
				<div>
					<img alt="none found" src={noImage} />
				</div>
			) : (
				<Cards columnCount={5}>
					{photos?.map((item) => (
						<Card key={item.id}>
							<Thumbnail
								imageSource={item.img_src}
								description={`${item.id}`}
							/>
						</Card>
					))}
				</Cards>
			)}
			<AlertStatus alertType="error" message={error}></AlertStatus>
		</div>
	);
}
