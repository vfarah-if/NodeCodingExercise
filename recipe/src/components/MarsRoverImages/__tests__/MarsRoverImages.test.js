import React from 'react';
import '@testing-library/jest-dom';
import { render, screen, waitFor } from '@testing-library/react';

import emptyResponse from './empty-response.json';
import curiosityFHAZResponse from './curiosity-fhaz-response.json';
import MarsRoverImages from '../MarsRoverImages';

describe('MarsRoverImages', () => {
	describe('When retrieving an empty response', () => {
		beforeEach(() => {
			jest.spyOn(window, 'fetch').mockImplementation(() =>
				Promise.resolve({
					json: () => Promise.resolve(emptyResponse),
				})
			);
		});

		test('should show success alert with zero images loaded', async () => {
			render(<MarsRoverImages roverType="spirit" cameraType="MAST" />);

			await waitFor(() => {
				expect(window.fetch).toHaveBeenCalledTimes(1);
				// console.debug(screen.debug());
				expect(
					screen.getByText(`Succeeded retrieving '0' photos!`)
				).toBeInTheDocument();
			});
		});

		test('should show not found image', async () => {
			render(<MarsRoverImages roverType="spirit" cameraType="MAST" />);

			await waitFor(() => {
				expect(window.fetch).toHaveBeenCalledTimes(1);
				// console.debug(screen.debug());
				expect(screen.getByAltText('none found')).toBeInTheDocument();
			});
		});
	});

	describe('When retrieving an full response', () => {
		beforeEach(() => {
			jest.spyOn(window, 'fetch').mockImplementation(() =>
				Promise.resolve({
					json: () => Promise.resolve(curiosityFHAZResponse),
				})
			);
		});

		test('should show success alert with zero images loaded', async () => {
			render(<MarsRoverImages roverType="curiosity" cameraType="FHA" />);

			await waitFor(() => {
				expect(window.fetch).toHaveBeenCalledTimes(1);
				// console.debug(screen.debug());
				expect(
					screen.getByText(`Succeeded retrieving '4' photos!`)
				).toBeInTheDocument();
			});

			expect(
				screen.getByText(`Succeeded retrieving '4' photos!`)
			).toMatchSnapshot();
		});

		test('should show one cardslist and four cards', async () => {
			const { container } = render(
				<MarsRoverImages roverType="curiosity" cameraType="FHA" />
			);

			await waitFor(() => {
				expect(window.fetch).toHaveBeenCalledTimes(1);
				// console.debug(screen.debug());
				expect(container.querySelectorAll('ul').length).toBe(1);
				expect(container.querySelectorAll('li').length).toBe(4);
			});

			expect(container.querySelector('ul')).toMatchSnapshot();
		});
	});

	describe('Select tests', () => {
		let selectArray, component;
		beforeEach(() => {
			jest.spyOn(window, 'fetch').mockImplementation(() =>
				Promise.resolve({
					json: () => Promise.resolve(emptyResponse),
				})
			);
		});

		test('should have two selects', async () => {
			component = render(
				<MarsRoverImages roverType="curiosity" cameraType="FHA" />
			);
			
			await waitFor(() => {
				selectArray = component.container.querySelectorAll('select');
				expect(selectArray.length).toBe(2);
			});

			expect(selectArray[0]).toMatchSnapshot('Rovers');
			expect(selectArray[1]).toMatchSnapshot('Cameras');
		});
	});
});
