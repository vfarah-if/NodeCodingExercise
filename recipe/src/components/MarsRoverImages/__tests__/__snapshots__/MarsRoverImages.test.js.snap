// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarsRoverImages Select tests should have first select setup with a list of rovers 1`] = `
<select
  class="select-css"
>
  <option
    value="curiosity"
  >
    Curiosity
  </option>
  <option
    value="spirit"
  >
    Spirit
  </option>
  <option
    value="opportunity"
  >
    Opportunity
  </option>
</select>
`;

exports[`MarsRoverImages Select tests should have second select setup with a list of cameras 1`] = `
<select
  class="select-css"
>
  <option
    value=""
  >
    Choose a camera
  </option>
  <optgroup
    label="Spirit or Opportunity or Curiosity"
  >
    <option
      value="NAVCAM"
    >
      Navigation Camera
    </option>
  </optgroup>
  <optgroup
    label="Spirit or Opportunity"
  >
    <option
      value="PANCAM"
    >
      Panoramic Camera
    </option>
  </optgroup>
  <optgroup
    label="Curiosity"
  >
    <option
      value="FHAZ"
    >
      Front Hazard Avoidance Camera
    </option>
    <option
      value="RHAZ"
    >
      Rear Hazard Avoidance Camera
    </option>
    <option
      value="MAST"
    >
      Mast Camera
    </option>
    <option
      value="CHEMCAM"
    >
      Chemistry and Camera Complex
    </option>
  </optgroup>
</select>
`;

exports[`MarsRoverImages When retrieving an full response should show one cardslist and four cards 1`] = `
<ul
  style="display: grid; grid-template-columns: repeat(5, 1fr);"
>
  <li
    class="card"
  >
    <img
      alt="669"
      class="thumbnail"
      src="http://mars.jpl.nasa.gov/msl-raw-images/proj/msl/redops/ods/surface/sol/00100/opgs/edr/fcam/FRA_406374643EDR_F0050178FHAZ00301M_.JPG"
    />
  </li>
  <li
    class="card"
  >
    <img
      alt="9487"
      class="thumbnail"
      src="http://mars.jpl.nasa.gov/msl-raw-images/proj/msl/redops/ods/surface/sol/00100/opgs/edr/fcam/FLA_406377515EDR_F0050178FHAZ00302M_.JPG"
    />
  </li>
  <li
    class="card"
  >
    <img
      alt="9488"
      class="thumbnail"
      src="http://mars.jpl.nasa.gov/msl-raw-images/proj/msl/redops/ods/surface/sol/00100/opgs/edr/fcam/FRA_406377515EDR_F0050178FHAZ00302M_.JPG"
    />
  </li>
  <li
    class="card"
  >
    <img
      alt="9489"
      class="thumbnail"
      src="http://mars.jpl.nasa.gov/msl-raw-images/proj/msl/redops/ods/surface/sol/00100/opgs/edr/fcam/FLA_406374643EDR_F0050178FHAZ00301M_.JPG"
    />
  </li>
</ul>
`;

exports[`MarsRoverImages When retrieving an full response should show success alert with zero images loaded 1`] = `
<div
  class="alert success"
>
  <span
    class="closebtn"
  >
    Ã—
  </span>
  Succeeded retrieving '4' photos!
</div>
`;
